src/app/ReadHandler.cpp

```diff
@@ -0,0 +1,238 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferHandle response;
+
+    mpExchangeCtx = apExchangeContext;
+
+    err = ProcessReadRequest(std::move(aPayload));
+    SuccessOrExit(err);
+
+exit:
+    ChipLogFunctError(err);
+    return;
+}
+
+CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kNone));
+    SuccessOrExit(err);
+    MoveToState(HandlerState::ReportDataSending);
+
+exit:
+    Shutdown();
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR ReadHandler::ProcessReadRequest(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVReader reader;
+
+    ReadRequest::Parser readRequestParser;
+    EventPathList::Parser eventPathListParser;
+    chip::TLV::TLVReader eventPathListReader;
+
+    reader.Init(std::move(aPayload));
+
+    err = reader.Next();
+    SuccessOrExit(err);
+
+    err = readRequestParser.Init(reader);
+    SuccessOrExit(err);
+
+    err = readRequestParser.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = readRequestParser.GetEventPathList(&eventPathListParser);
+    SuccessOrExit(err);
+
+    eventPathListParser.GetReader(&eventPathListReader);
+
+    while (CHIP_NO_ERROR == (err = eventPathListReader.Next()))
+    {
+        VerifyOrExit(chip::TLV::AnonymousTag == eventPathListReader.GetTag(), err = CHIP_ERROR_INVALID_TLV_TAG);
+
+        EventPath::Parser eventPath;
+
+        err = eventPath.Init(eventPathListReader);
+        SuccessOrExit(err);
+
+        // TODO: Process and offer EventPath for RE to filter interested path
+    }
+
+    // if we have exhausted this container
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+
+    InteractionModelEngine::GetInstance()->GetReportingEngine()->ScheduleRun();
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+const char * ReadHandler::GetStateStr() const
+{
+#if CHIP_DETAIL_LOGGING
+    switch (mState)
+    {
+    case HandlerState::Uninitialized:
+        return "Uninitialized";
+
+    case HandlerState::Initialized:
+        return "Initialized";
+
+    case HandlerState::ReportDataSending:
+        return "ReportDataSending";
+    }
+#endif // CHIP_DETAIL_LOGGING
+    return "N/A";
+}
+
+void ReadHandler::MoveToState(const HandlerState aTargetState)
+{
+    mState = aTargetState;
+    ChipLogDetail(DataManagement, "IM RH moving to [%10.10s]", GetStateStr());
+}
+
+void ReadHandler::ClearState(void)
+{
+    MoveToState(HandlerState::Uninitialized);
+}
+
+bool ReadHandler::CheckEventUpToDate(reporting::LoggingManagement & inLogger)
+{
+    bool retval = true;
```

Comment:
please just return true/false when relevant. The file would be so much clearer:

```cpp

if (!inLogger.IsValid()) {
   return true;
}

for (uint8_t i = 0; i < ArraySize(mSelfVendedEvents); i++)
{
    chip::EventNumber eid = inLogger.GetLastEventNumber(
                static_cast<reporting::PriorityLevel>(i + static_cast<uint8_t>(reporting::PriorityLevel::First)));
    if ((eid != 0) && (eid >= mSelfVendedEvents[i]))
    {
        return false;
    }
}
        
return true;
```

this applies throught (e.g. for FindNextPriorityForTransfer)